library/std/src/panic.rs

```diff
@@ -285,45 +285,54 @@ where
 /// It is **not** recommended to use this function for a general try/catch
 /// mechanism. The [`Result`] type is more appropriate to use for functions that
 /// can fail on a regular basis. Additionally, this function is not guaranteed
 /// to catch all panics, see the "Notes" section below.
 ///
-/// The closure provided is required to adhere to the [`UnwindSafe`] trait to ensure
-/// that all captured variables are safe to cross this boundary. The purpose of
-/// this bound is to encode the concept of [exception safety][rfc] in the type
-/// system. Most usage of this function should not need to worry about this
-/// bound as programs are naturally unwind safe without `unsafe` code. If it
-/// becomes a problem the [`AssertUnwindSafe`] wrapper struct can be used to quickly
-/// assert that the usage here is indeed unwind safe.
+/// The closure provided is required to adhere to the [`UnwindSafe`] trait to
+/// ensure that all captured variables are safe to cross this boundary. The
+/// purpose of this bound is to encode the concept of [exception safety][rfc] in
+/// the type system. Most usage of this function should not need to worry about
+/// this bound as programs are naturally unwind safe without `unsafe` code. If
+/// it becomes a problem the [`AssertUnwindSafe`] wrapper struct can be used to
+/// quickly assert that the usage here is indeed unwind safe.
 ///
 /// [rfc]: https://github.com/rust-lang/rfcs/blob/master/text/1236-stabilize-catch-panic.md
 ///
 /// # Notes
 ///
-/// Note that this function **might not catch all panics** in Rust. A panic in
-/// Rust is not always implemented via unwinding, but can be implemented by
-/// aborting the process as well. This function *only* catches unwinding panics,
-/// not those that abort the process.
+/// This function **might not catch all Rust panics**. A Rust panic is not
+/// always implemented via unwinding, but can be implemented by aborting the
+/// process as well. This function *only* catches unwinding panics, not those
+/// that abort the process.
 ///
-/// Note that if a custom panic hook has been set, it will be invoked before
-/// the panic is caught, before unwinding.
+/// If a custom panic hook has been set, it will be invoked before the panic is
+/// caught, before unwinding.
 ///
-/// Also note that unwinding into Rust code with a foreign exception (e.g.
-/// an exception thrown from C++ code) is undefined behavior.
+/// Although unwinding into Rust code with a foreign exception (e.g. an
+/// exception thrown from C++ code, or a `panic!` in Rust code compiled or
+/// linked with a different runtime) via an appropriate ABI (e.g. `"C-unwind"`)
+/// is permitted, catching such an exception using this function will have one
+/// of two behaviors, and it is unspecified which will occur:
 ///
-/// Finally, be **careful in how you drop the result of this function**.
-/// If it is `Err`, it contains the panic payload, and dropping that may in turn panic!
+/// * The process aborts.
```

Comment:
It would be good to explicitly state that the abort happens after unwinding out of the closure / into the `catch_unwind` frame (i.e. it would be incorrect to abort from the personality routine).

Reply:
Isn't this related to https://github.com/rust-lang/rust/issues/123231, where aborts may be happening earlier than expected?

Reply:
Yes, it's directly related. But the scenarios discussed in this documentation should be more straightforward than the fully general case, if I recall the details with any reasonable accuracy. (Because we control the catch frame involved in the discussed cases.)

Reply:
The fact that the closure itself can't abort is also implied by the fact that we want the standard library to have control over the selected behavior here, right?

Reply:
I don't know, this abort-early-on-unwind thing is entirely mysterious to me. Cc @nbdd0121

Reply:
A pre-unwinding abort could happen in one of two cases with Itanium EH, AIUI:

- The personality routine (what selects the unwinding behavior for a frame) directly aborts in phase 1.
- No unwinding handler ("catch frame") is present, causing the unwind to fail to throw. 

This is a consequence of Itanium EH's two phase lookup.

Reply:
I guess we could in theory implement a personality routine which fails phase 1 if `catch_unwind` is about to catch a foreign exception. Not sure how much value would it provide though.

Reply:
@nbdd0121 @CAD97 @RalfJung 

Do we need to make a change here before merge, or not? I'm afraid I'm not entirely following the implications of this comment thread.

Reply:
I think @CAD97 asked for this to say something like
"The process aborts, after executing all destructors of `f` and the functions it called."

Reply:
Added that verbiage.